app/controllers/MovementsController.scala

```diff
@@ -20,87 +20,166 @@ import java.time.OffsetDateTime
-                                // TODO: This needs to be replaced url to arrival movement resource, for which we need an Arrival Movement number
-                                  .withHeaders("Location" -> arrival.arrivalId.index.toString)
-
-                            }
-                            .recoverWith {
-                              case e: Exception =>
-                                logger.error(s"setState failed with following Exception: ${e.getMessage}")
-                                Future.successful(InternalServerError)
-                            }
-                      }
-                      .recoverWith {
-                        case error: Exception =>
-                          logger.error(s"Call to EIS failed with the following Exception: ${error.getMessage}")
-                          val newState = arrival.state.transition(SubmissionResult.Failure)
-                          arrivalMovementRepository.setState(arrival.arrivalId, newState).map {
+          arrivalMovementService.makeArrivalMovement(request.eoriNumber)(request.body) match {
+            case None =>
+              Future.successful(BadRequest("Invalid data: missing either DatOfPreMES9, TimOfPreMES10 or DocNumHEA5"))
+            case Some(x) =>
+              x.flatMap {
+                  arrival =>
+                    arrivalMovementRepository.insert(arrival) flatMap {
+                      _ =>
+                        messageConnector
+                          .post(
+                            TransitWrapper(request.body),
+                            MessageType.ArrivalNotification,
+                            OffsetDateTime.now,
+                            MessageSender(arrival.arrivalId, arrival.messages.head.messageCorrelationId)
+                          )
+                          .flatMap {
                             _ =>
-                              BadGateway
+                              arrival.messages.last.getState() match {
+                                case Some(state) =>
+                                  for {
+                                    _ <- arrivalMovementRepository.setMessageState(arrival.arrivalId,
+                                                                                   arrival.messages.length - 1,
+                                                                                   state.transition(SubmissionResult.Success))
+                                    _ <- arrivalMovementRepository.setState(arrival.arrivalId, arrival.state.transition(ArrivalSubmitted))
+                                  } yield {
+                                    Accepted("Message accepted")
+                                      .withHeaders("Location" -> routes.MovementsController.get(arrival.arrivalId).url)
+                                  }
+                                case None =>
+                                  Future.successful(InternalServerError)
```

Comment:
Consider pulling this out into it's own method so it's clear it's setting the movement and message state

Reply:
I'll be adding a way to do these both in a single write to mongo so that they get the benefit of being atomic